examples/fa-test.cpp

```diff
@@ -0,0 +1,806 @@
+   }
+   if (mesh->NURBSext)
+   {
+      mesh->SetCurvature(max(order, 1));
+   }
+   mesh->GetBoundingBox(bb_min, bb_max, max(order, 1));
+
+   // 6. Define the parallel mesh by a partitioning of the serial mesh. Refine
+   //    this mesh further in parallel to increase the resolution. Once the
+   //    parallel mesh is defined, the serial mesh can be deleted.
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int lev = 0; lev < par_ref_levels; lev++)
+   {
+      pmesh->UniformRefinement();
+   }
+
+   // 7. Define the parallel discontinuous DG finite element space on the
+   //    parallel refined mesh of the given polynomial order.
+   DG_FECollection fec(order, dim, BasisType::GaussLobatto);
+   ParFiniteElementSpace *fes = new ParFiniteElementSpace(pmesh, &fec);
+
+   HYPRE_Int global_vSize = fes->GlobalTrueVSize();
+   if (myid == 0)
+   {
+      cout << "Number of unknowns: " << global_vSize << endl;
+   }
+
+   // 8. Set up and assemble the parallel bilinear and linear forms (and the
+   //    parallel hypre matrices) corresponding to the DG discretization. The
+   //    DGTraceIntegrator involves integrals over mesh interior faces.
+   VectorFunctionCoefficient velocity(dim, velocity_function);
+   FunctionCoefficient inflow(inflow_function);
+   FunctionCoefficient u0(u0_function);
+
+   ParBilinearForm *m = new ParBilinearForm(fes);
+   ParBilinearForm *k = new ParBilinearForm(fes);
+   if (pa)
+   {
+      m->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      k->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   }
+   else if (ea)
+   {
+      m->SetAssemblyLevel(AssemblyLevel::ELEMENT);
+      k->SetAssemblyLevel(AssemblyLevel::ELEMENT);
+   }
+   else if (fa)
+   {
+      m->SetAssemblyLevel(AssemblyLevel::FULL);
+      k->SetAssemblyLevel(AssemblyLevel::FULL);
+   }
+
+   m->AddDomainIntegrator(new MassIntegrator);
+   k->AddDomainIntegrator(new ConvectionIntegrator(velocity, -1.0));
+   k->AddInteriorFaceIntegrator(
+      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
+   //k->AddBdrFaceIntegrator(
+   //new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
+
+   ParLinearForm *b = new ParLinearForm(fes);
+   b->AddBdrFaceIntegrator(
+      new BoundaryFlowIntegrator(inflow, velocity, -1.0, -0.5));
+
+   int skip_zeros = 0;
+   m->Assemble();
+   k->Assemble(skip_zeros);
+   b->Assemble();
+   m->Finalize();
+   k->Finalize(skip_zeros);
+
+   k->KeepNbrBlock();
+
+   SparseMatrix A = GetFullAssemblySparseMatrix(*k);
+   
+   //GetFullAssemblySparseMatrix(k, A);
+
+   //Reference code -- legacy assembly 
+   ParBilinearForm *k_ref = new ParBilinearForm(fes);
+   k_ref->AddDomainIntegrator(new ConvectionIntegrator(velocity, -1.0));
+   k_ref->AddInteriorFaceIntegrator(
+      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
+   
+   k_ref->Finalize(skip_zeros);
+   k_ref->KeepNbrBlock();
+
+   double iError(0), jError(0), dataError(0);
```

Comment:
@YohannDudouit , one way to check if the routine works is to compare the CSR data structure to the legacy version of the code (no boundary conditions imposed here). Below I check the I, J, Data arrays. Currently this example  segfaults, but getting this to work would be a good starting point. Ideally, we could then reuse the sparse matrix to repopulate on the GPU - for time dependent coefficient.

Reply:
@artv3 The test you wrote is SegFaulting in something that has nothing to do with `GetFullAssemblySparseMatrix`, can you fix it?

Reply:
Yes I can! Fixed and Pushed :-). 
I added some print statements to look at the sizes of the array. I see some differences between J, and Data: 
 lrun -n 1 ./fa-test gives 
```
size of I k_ref 3073 A 3073 
size of J k_ref 245760 A 61440 
size of Data k_ref 245760 A 61440 
```

Reply:
I think your test doesn't really work, in this example the `SparseMatrix` are the same but your error is high:
```
 ~/Projects/mfem/examples/> ./fa-test -fa -m ../data/inline-quad.mesh -o 1
Options used:
   --mesh ../data/inline-quad.mesh
   --problem 0
   --refine-serial 0
   --refine-parallel 0
   --order 1
   --no-partial-assembly
   --no-element-assembly
   --full-assembly
   --device cpu
   --ode-solver 4
   --t-final 10
   --time-step 0.01
   --visualization
   --no-visit-datafiles
   --no-paraview-datafiles
   --no-adios2-streams
   --ascii-datafiles
   --visualization-steps 5
Device configuration: cpu
Memory configuration: host-std
Number of unknowns: 4
A=
[row 0]
 (0,0.23230781) (1,-0.087970241) (2,-0.028183663) (3,-0.1161539)
[row 1]
 (0,0.18419529) (1,-0.039857719) (2,0.019928859) (3,-0.16426643)
[row 2]
 (0,0.16426643) (1,-0.019928859) (2,0.039857719) (3,-0.18419529)
[row 3]
 (0,0.1161539) (1,0.028183663) (2,0.087970241) (3,-0.23230781)

A_ref=
[row 0]
 (3,-0.1161539) (2,-0.028183663) (1,-0.087970241) (0,0.23230781)
[row 1]
 (3,-0.16426643) (2,0.019928859) (1,-0.039857719) (0,0.18419529)
[row 2]
 (3,-0.18419529) (2,0.039857719) (1,-0.019928859) (0,0.16426643)
[row 3]
 (3,-0.23230781) (2,0.087970241) (1,0.028183663) (0,0.1161539)

size of I k_ref 5 A 5 
size of J k_ref 16 A 16 
size of Data k_ref 16 A 16 
iError 0 
jError 8.94427 
dataError 1 
error too high 
```

Reply:
Looking at a 2 elements mesh, the matrices are still the same, but the legacy sparse matrix seems to be storing a lot of zeros:
```
 ~/Projects/mfem/examples/> ./fa-test -fa -m ../data/inline-quad.mesh -o 1
Options used:
   --mesh ../data/inline-quad.mesh
   --problem 0
   --refine-serial 0
   --refine-parallel 0
   --order 1
   --no-partial-assembly
   --no-element-assembly
   --full-assembly
   --device cpu
   --ode-solver 4
   --t-final 10
   --time-step 0.01
   --visualization
   --no-visit-datafiles
   --no-paraview-datafiles
   --no-adios2-streams
   --ascii-datafiles
   --visualization-steps 5
Device configuration: cpu
Memory configuration: host-std
Number of unknowns: 8
A=
[row 0]
 (0,0.18419529) (1,-0.1120265) (2,0.019928859) (3,-0.092097643)
[row 1]
 (0,0.16013902) (1,-0.087970241) (2,0.043985121) (3,-0.1161539)
 (4,0) (6,0)
[row 2]
 (0,0.1161539) (1,-0.043985121) (2,0.087970241) (3,-0.16013902)
[row 3]
 (0,0.092097643) (1,-0.019928859) (2,0.1120265) (3,-0.18419529)
 (4,0) (6,0)
[row 4]
 (4,-0.087970241) (5,-0.1120265) (6,-0.1161539) (7,-0.092097643)
 (1,0.27216553) (3,0.13608276)
[row 5]
 (4,0.16013902) (5,-0.087970241) (6,0.043985121) (7,-0.1161539)
[row 6]
 (4,-0.019928859) (5,-0.043985121) (6,-0.18419529) (7,-0.16013902)
 (1,0.13608276) (3,0.27216553)
[row 7]
 (4,0.092097643) (5,-0.019928859) (6,0.1120265) (7,-0.18419529)

A_ref=
[row 0]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.092097643) (2,0.019928859) (1,-0.1120265) (0,0.18419529)
[row 1]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.1161539) (2,0.043985121) (1,-0.087970241) (0,0.16013902)
[row 2]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.16013902) (2,0.087970241) (1,-0.043985121) (0,0.1161539)
[row 3]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.18419529) (2,0.1120265) (1,-0.019928859) (0,0.092097643)
[row 4]
 (3,0.13608276) (2,0) (1,0.27216553) (0,0)
 (7,-0.092097643) (6,-0.1161539) (5,-0.1120265) (4,-0.087970241)
[row 5]
 (3,0) (2,0) (1,0) (0,0)
 (7,-0.1161539) (6,0.043985121) (5,-0.087970241) (4,0.16013902)
[row 6]
 (3,0.27216553) (2,0) (1,0.13608276) (0,0)
 (7,-0.16013902) (6,-0.18419529) (5,-0.043985121) (4,-0.019928859)
[row 7]
 (3,0) (2,0) (1,0) (0,0)
 (7,-0.18419529) (6,0.1120265) (5,-0.019928859) (4,0.092097643)

size of I k_ref 9 A 9 
size of J k_ref 64 A 40 
size of Data k_ref 64 A 40 
iError 42.332 
jError nan 
dataError 1.17465 
error too high 
```

Reply:
I also added:
```
   k->AddBdrFaceIntegrator(
      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
```
and the `SparseMatrix` are still the same.

Reply:
(it's only if essential dofs, I think, that the results would be different)

Reply:
Interesting (I forgot they might be storing a different number of zeros -- although I was hoping in CSR we wouldn't store zeros... ) ! I turned on skip_zeros (skip_zeros=1) in my last push, and I did see some differences in array sizes. But if they are still the same matrices but with different numbers of zeros in it, then that is fine. We can just do a convergence test with a periodic mesh and if it checks out we can confirm it is correct.

Reply:
Also, I just noticed the column indices are sorted in the new implementation, so of course they won't agree unless we sort them...

Reply:
Idea! Perhaps we can filter out zeros and sort the columns when comparing the old and new matrices.

Reply:
I don't really sort them, I happen to insert the values in order, this might not hold on GPU, due to concurrent access.

Reply:
I'm not sure this will be true for meshes with more elements.

Reply:
It's actually already not holding:
```
[row 6]
 (4,-0.019928859) (5,-0.043985121) (6,-0.18419529) (7,-0.16013902)
 (1,0.13608276) (3,0.27216553)
```

Reply:
You would need to sort both of them.